File: android/app/src/main/java/com/oppzippy/openscq30/lib/wrapper/CustomButtonModel.kt
package com.oppzippy.openscq30.lib.wrapper

import com.oppzippy.openscq30.R
import com.oppzippy.openscq30.lib.protobuf.customButtonModel
import com.oppzippy.openscq30.lib.protobuf.noTwsButtonAction
import com.oppzippy.openscq30.lib.protobuf.twsButtonAction

data class CustomButtonModel(
    val leftSingleClick: NoTwsButtonAction,
    val leftDoubleClick: TwsButtonAction,
    val leftLongPress: TwsButtonAction,
    val rightSingleClick: NoTwsButtonAction,
    val rightDoubleClick: TwsButtonAction,
    val rightLongPress: TwsButtonAction,
) {
    fun toProtobuf(): com.oppzippy.openscq30.lib.protobuf.CustomButtonModel = customButtonModel {
        leftSingleClick = this@CustomButtonModel.leftSingleClick.toProtobuf()
        leftDoubleClick = this@CustomButtonModel.leftDoubleClick.toProtobuf()
        leftLongPress = this@CustomButtonModel.leftLongPress.toProtobuf()
        rightSingleClick = this@CustomButtonModel.rightSingleClick.toProtobuf()
        rightDoubleClick = this@CustomButtonModel.rightDoubleClick.toProtobuf()
        rightLongPress = this@CustomButtonModel.rightLongPress.toProtobuf()
    }
}

fun com.oppzippy.openscq30.lib.protobuf.CustomButtonModel.toKotlin(): CustomButtonModel = CustomButtonModel(
    leftSingleClick = leftSingleClick.toKotlin(),
    leftDoubleClick = leftDoubleClick.toKotlin(),
    leftLongPress = leftLongPress.toKotlin(),
    rightSingleClick = rightSingleClick.toKotlin(),
    rightDoubleClick = rightDoubleClick.toKotlin(),
    rightLongPress = rightLongPress.toKotlin(),
)

data class TwsButtonAction(
    val isEnabled: Boolean,
    val twsConnectedAction: ButtonAction,
    val twsDisconnectedAction: ButtonAction,
) {
    fun toProtobuf(): com.oppzippy.openscq30.lib.protobuf.TwsButtonAction = twsButtonAction {
        isEnabled = this@TwsButtonAction.isEnabled
        twsConnectedAction = this@TwsButtonAction.twsDisconnectedAction.toProtobuf()
        twsDisconnectedAction = this@TwsButtonAction.twsDisconnectedAction.toProtobuf()
    }

    fun connectedActionOrNull(): ButtonAction? = if (isEnabled) {
        twsConnectedAction
    } else {
        null
    }
}

fun com.oppzippy.openscq30.lib.protobuf.TwsButtonAction.toKotlin(): TwsButtonAction = TwsButtonAction(
    isEnabled = isEnabled,
    twsConnectedAction = twsConnectedAction.toKotlin(),
    twsDisconnectedAction = twsDisconnectedAction.toKotlin(),
)

data class NoTwsButtonAction(val isEnabled: Boolean, val action: ButtonAction) {
    fun toProtobuf(): com.oppzippy.openscq30.lib.protobuf.NoTwsButtonAction = noTwsButtonAction {
        isEnabled = this@NoTwsButtonAction.isEnabled
        action = this@NoTwsButtonAction.action.toProtobuf()
    }

    fun actionOrNull(): ButtonAction? = if (isEnabled) {
        action
    } else {
        null
    }
}

fun com.oppzippy.openscq30.lib.protobuf.NoTwsButtonAction.toKotlin(): NoTwsButtonAction =
    NoTwsButtonAction(isEnabled, action.toKotlin())

enum class ButtonAction {
    VolumeUp,
    VolumeDown,
    PreviousSong,
    NextSong,
    AmbientSoundMode,
    VoiceAssistant,
    PlayPause,
    GameMode,
    ;

    fun toProtobuf(): com.oppzippy.openscq30.lib.protobuf.ButtonAction = when (this) {
        VolumeUp -> com.oppzippy.openscq30.lib.protobuf.ButtonAction.VOLUME_UP
        VolumeDown -> com.oppzippy.openscq30.lib.protobuf.ButtonAction.VOLUME_DOWN
        PreviousSong -> com.oppzippy.openscq30.lib.protobuf.ButtonAction.PREVIOUS_SONG
        NextSong -> com.oppzippy.openscq30.lib.protobuf.ButtonAction.NEXT_SONG
        AmbientSoundMode -> com.oppzippy.openscq30.lib.protobuf.ButtonAction.AMBIENT_SOUND_MODE
        VoiceAssistant -> com.oppzippy.openscq30.lib.protobuf.ButtonAction.VOICE_ASSISTANT
        PlayPause -> com.oppzippy.openscq30.lib.protobuf.ButtonAction.PLAY_PAUSE
        GameMode -> com.oppzippy.openscq30.lib.protobuf.ButtonAction.GAME_MODE
    }

    fun toStringResource(): Int = when (this) {
        VolumeUp -> R.string.volume_up
        VolumeDown -> R.string.volume_down
        PreviousSong -> R.string.previous_song
        NextSong -> R.string.next_song
        AmbientSoundMode -> R.string.ambient_sound_mode
        VoiceAssistant -> R.string.voice_assistant
        PlayPause -> R.string.play_pause
        GameMode -> R.string.game_mode
    }
}

fun com.oppzippy.openscq30.lib.protobuf.ButtonAction.toKotlin(): ButtonAction = when (this) {
    com.oppzippy.openscq30.lib.protobuf.ButtonAction.VOLUME_UP -> ButtonAction.VolumeUp
    com.oppzippy.openscq30.lib.protobuf.ButtonAction.VOLUME_DOWN -> ButtonAction.VolumeDown
    com.oppzippy.openscq30.lib.protobuf.ButtonAction.PREVIOUS_SONG -> ButtonAction.PreviousSong
    com.oppzippy.openscq30.lib.protobuf.ButtonAction.NEXT_SONG -> ButtonAction.NextSong
    com.oppzippy.openscq30.lib.protobuf.ButtonAction.AMBIENT_SOUND_MODE -> ButtonAction.AmbientSoundMode
    com.oppzippy.openscq30.lib.protobuf.ButtonAction.VOICE_ASSISTANT -> ButtonAction.VoiceAssistant
    com.oppzippy.openscq30.lib.protobuf.ButtonAction.PLAY_PAUSE -> ButtonAction.PlayPause
    com.oppzippy.openscq30.lib.protobuf.ButtonAction.GAME_MODE -> ButtonAction.GameMode
}

File: lib/src/devices/standard/structures/custom_button_model.rs
use nom::{
    combinator::{map, map_opt},
    error::{context, ContextError, ParseError},
    number::complete::le_u8,
    sequence::{pair, tuple},
    IResult,
};
#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};
use strum::{AsRefStr, EnumIter, FromRepr};

use crate::devices::standard::packets::parsing::take_bool;

#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
pub struct CustomButtonActions {
    pub left_single_click: ButtonState,
    pub left_double_click: ButtonState,
    pub left_long_press: ButtonState,
    pub right_single_click: ButtonState,
    pub right_double_click: ButtonState,
    pub right_long_press: ButtonState,
}

#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
pub struct ButtonState {
    pub action: ButtonAction,
    pub is_enabled: bool,
}

#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
pub(crate) struct InternalCustomButtonModel {
    pub left_single_click: NoTwsButtonAction,
    pub left_double_click: TwsButtonAction,
    pub left_long_press: TwsButtonAction,
    pub right_single_click: NoTwsButtonAction,
    pub right_double_click: TwsButtonAction,
    pub right_long_press: TwsButtonAction,
}

impl From<InternalCustomButtonModel> for CustomButtonActions {
    fn from(value: InternalCustomButtonModel) -> Self {
        Self {
            left_single_click: value.left_single_click.into(),
            left_double_click: value.left_double_click.into(),
            left_long_press: value.left_long_press.into(),
            right_single_click: value.right_single_click.into(),
            right_double_click: value.right_double_click.into(),
            right_long_press: value.right_long_press.into(),
        }
    }
}

impl InternalCustomButtonModel {
    pub fn bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::with_capacity(12);
        bytes.extend(self.left_double_click.bytes());
        bytes.extend(self.left_long_press.bytes());
        bytes.extend(self.right_double_click.bytes());
        bytes.extend(self.right_long_press.bytes());
        bytes.extend(self.left_single_click.bytes());
        bytes.extend(self.right_single_click.bytes());
        bytes
    }

    pub(crate) fn take<'a, E: ParseError<&'a [u8]> + ContextError<&'a [u8]>>(
        input: &'a [u8],
    ) -> IResult<&'a [u8], InternalCustomButtonModel, E> {
        context("custom button model", |input| {
            map(
                tuple((
                    context("left double click", TwsButtonAction::take), // left double click
                    context("left long press", TwsButtonAction::take),   // left long press
                    context("right double click", TwsButtonAction::take), // right double click
                    context("right long press", TwsButtonAction::take),  // right long press
                    context("left single click", NoTwsButtonAction::take), // left single click
                    context("right single click", NoTwsButtonAction::take), // right single click
                )),
                |(
                    left_double_click,
                    left_long_press,
                    right_double_click,
                    right_long_press,
                    left_single_press,
                    right_single_press,
                )| {
                    InternalCustomButtonModel {
                        left_double_click,
                        left_long_press,
                        right_double_click,
                        right_long_press,
                        left_single_click: left_single_press,
                        right_single_click: right_single_press,
                    }
                },
            )(input)
        })(input)
    }
}

#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
pub(crate) struct TwsButtonAction {
    pub tws_connected_action: ButtonAction,
    pub tws_disconnected_action: ButtonAction,
    pub disconnected_switch: bool,
}

impl From<TwsButtonAction> for ButtonState {
    fn from(value: TwsButtonAction) -> Self {
        Self {
            action: value.active_action(),
            is_enabled: true,
        }
    }
}

impl TwsButtonAction {
    pub fn bytes(&self) -> [u8; 2] {
        [
            self.disconnected_switch.into(),
            (u8::from(self.tws_disconnected_action) << 4)
                | (u8::from(self.tws_connected_action) & 0x0f),
        ]
    }

    pub(crate) fn take<'a, E: ParseError<&'a [u8]> + ContextError<&'a [u8]>>(
        input: &'a [u8],
    ) -> IResult<&'a [u8], TwsButtonAction, E> {
        map_opt(pair(take_bool, le_u8), |(switch, num)| {
            Some(TwsButtonAction {
                tws_connected_action: ButtonAction::from_repr(num & 0x0F)?,
                tws_disconnected_action: ButtonAction::from_repr((num & 0xF0) >> 4)?,
                disconnected_switch: switch,
            })
        })(input)
    }

    pub fn set_action(&mut self, action: ButtonAction, is_tws_connected: bool) {
        self.disconnected_switch = !is_tws_connected;
        if self.disconnected_switch {
            self.tws_disconnected_action = action;
        } else {
            self.tws_connected_action = action;
        }
    }

    pub fn active_action(&self) -> ButtonAction {
        if self.disconnected_switch {
            self.tws_disconnected_action
        } else {
            self.tws_connected_action
        }
    }
}

#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
pub(crate) struct NoTwsButtonAction {
    pub action: ButtonAction,
    pub is_enabled: bool,
}

impl From<NoTwsButtonAction> for ButtonState {
    fn from(value: NoTwsButtonAction) -> Self {
        Self {
            action: value.action,
            is_enabled: value.is_enabled,
        }
    }
}

impl NoTwsButtonAction {
    pub fn bytes(&self) -> [u8; 2] {
        [self.is_enabled.into(), u8::from(self.action) & 0x0f]
    }

    pub(crate) fn take<'a, E: ParseError<&'a [u8]> + ContextError<&'a [u8]>>(
        input: &'a [u8],
    ) -> IResult<&'a [u8], NoTwsButtonAction, E> {
        map_opt(pair(take_bool, le_u8), |(switch, num)| {
            Some(NoTwsButtonAction {
                action: ButtonAction::from_repr(num)?,
                is_enabled: switch,
            })
        })(input)
    }
}

#[derive(
    Default, Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord, FromRepr, EnumIter, AsRefStr,
)]
#[repr(u8)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
pub enum ButtonAction {
    #[default]
    VolumeUp = 0,
    VolumeDown = 1,
    PreviousSong = 2,
    NextSong = 3,
    AmbientSoundMode = 4,
    VoiceAssistant = 5,
    PlayPause = 6,
    GameMode = 9,
}

impl From<ButtonAction> for u8 {
    fn from(value: ButtonAction) -> Self {
        value as Self
    }
}
